src/Validation/ScannableURLsRestController.php

```diff
@@ -0,0 +1,139 @@
+<?php
+/**
+ * ScannableURLsRestController class.
+ *
+ * @package AmpProject\AmpWP
+ * @since 2.2
+ */
+
+namespace AmpProject\AmpWP\Validation;
+
+use AMP_Validation_Manager;
+use WP_Error;
+use WP_REST_Controller;
+use AmpProject\AmpWP\Infrastructure\Delayed;
+use AmpProject\AmpWP\Infrastructure\Registerable;
+use AmpProject\AmpWP\Infrastructure\Service;
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * Rest endpoint for fetching the scannable URLs.
+ *
+ * @since 2.2
+ * @internal
+ */
+final class ScannableURLsRestController extends WP_REST_Controller implements Delayed, Service, Registerable {
+
+	/** @var ScannableURLProvider ScannableURLProvider instance. */
+	private $scannable_url_provider;
+
+	/**
+	 * Get the action to use for registering the service.
+	 *
+	 * @return string Registration action to use.
+	 */
+	public static function get_registration_action() {
+		return 'rest_api_init';
+	}
+
+	/**
+	 * Constructor.
+	 *
+	 * @param ScannableURLProvider $scannable_url_provider Instance of ScannableURLProvider class.
+	 */
+	public function __construct( ScannableURLProvider $scannable_url_provider ) {
+		$this->namespace              = 'amp/v1';
+		$this->rest_base              = 'scannable-urls';
+		$this->scannable_url_provider = $scannable_url_provider;
+	}
+
+	/**
+	 * Registers all routes for the controller.
+	 */
+	public function register() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'get_items' ],
+					'args'                => [],
+					'permission_callback' => static function () {
+						return current_user_can( \AMP_Validation_Manager::VALIDATE_CAPABILITY );
+					},
+				],
+				'schema' => [ $this, 'get_public_item_schema' ],
+			]
+		);
+	}
+
+	/**
+	 * Retrieves a list of scannable URLs.
+	 *
+	 * Each item contains a page `type` (e.g. 'home' or 'search') and a
+	 * `validate_url` prop for accessing validation data for a given URL.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		$nonce = AMP_Validation_Manager::get_amp_validate_nonce();
+
+		return rest_ensure_response(
+			array_map(
+				static function ( $entry ) use ( $nonce ) {
+					$entry['validate_url'] = amp_add_paired_endpoint(
```

Comment:
I guess that once it's possible to [temporarily override template mode](https://github.com/ampproject/amp-wp/pull/6615/files#r713858260), there will be no need for using `amp_add_paired_endpoint` method here.

In fact, it seems that `amp_add_paired_endpoint` adds an AMP query parameter even if Standard mode is selected resulting with 302 redirects when validating:

![Screenshot 2021-09-22 at 13 56 59](https://user-images.githubusercontent.com/478735/134339303-2bd3ab51-de75-46f0-8fef-abd32d8c0dd3.png)

Reply:
Yes, good point! Instead of adding the paired endpoint, this should rather add the parameter to force AMP.

Reply:
I'm wondering if the `amp_validate` param could do that (i.e. force AMP in standard mode)? After all, it's impossible to validate a non-AMP page so having the `amp_validate` param in the request should be enough, correct?

Reply:
Ironically, this used to be the case before v2.1:

https://github.com/ampproject/amp-wp/blob/f9c514575b3c15fa009257e2f449ea2ee03d5700/includes/amp-helper-functions.php#L606-L614

It was removed to fix #5295.

But we can bring it back with an additional parameter to force AMP to be available. That way it will only be be overridden if explicitly requested.

Reply:
> But we can bring it back with an additional parameter to force AMP to be available. That way it will only be be overridden if explicitly requested.

See https://github.com/ampproject/amp-wp/pull/6615